lib/plug/adapters/cowboy2/handler.ex

```diff
@@ -0,0 +1,58 @@
+defmodule Plug.Adapters.Cowboy2.Handler do
+  @moduledoc false
+  @connection Plug.Adapters.Cowboy2.Conn
+  @already_sent {:plug_conn, :sent}
+
+  def init(req, {plug, opts}) do
+    {__MODULE__, req, {plug, opts}}
+  end
+
+  def upgrade(req, env, __MODULE__, {plug, opts}) do
+    conn = @connection.conn(req)
+
+    try do
+      %{adapter: {@connection, req}} =
+        conn
+        |> plug.call(opts)
+        |> maybe_send(plug)
+
+      {:ok, req, Map.put_new(env, :result, :ok)}
+    catch
+      :error, value ->
+        stack = System.stacktrace()
+        exception = Exception.normalize(:error, value, stack)
+        reason = {{exception, stack}, {plug, :call, [conn, opts]}}
+        terminate(reason, req, stack)
+
+      :throw, value ->
+        stack = System.stacktrace()
+        reason = {{{:nocatch, value}, stack}, {plug, :call, [conn, opts]}}
+        terminate(reason, req, stack)
+
+      :exit, value ->
+        stack = System.stacktrace()
+        reason = {value, {plug, :call, [conn, opts]}}
+        terminate(reason, req, stack)
+    after
+      receive do
+        @already_sent -> :ok
+      after
+        0 -> :ok
+      end
+    end
+  end
+
+  defp maybe_send(%Plug.Conn{state: :unset}, _plug), do: raise(Plug.Conn.NotSentError)
+  defp maybe_send(%Plug.Conn{state: :set} = conn, _plug), do: Plug.Conn.send_resp(conn)
+  defp maybe_send(%Plug.Conn{} = conn, _plug), do: conn
+
+  defp maybe_send(other, plug) do
+    raise "Cowboy2 adapter expected #{inspect(plug)} to return Plug.Conn but got: #{
```

Comment:
We need to break this into using `<>` to avoid this weird interpolation thing. :)